### ISSUE TITLE
apt_key module fails to download the key from the URL

### ISSUE THREAD

**Issue by tadeboro** [ansible/ansible#70998]
##### SUMMARY
apt_key module fails to download keys from username/password protected address if response from that address redirects to a different url

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
apt_key

##### ANSIBLE VERSION
```
ansible 2.9.10
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/tadej/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.8/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.8.3 (default, May 29 2020, 00:00:00) [GCC 10.1.1 20200507 (Red Hat 10.1.1-1)]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT
Ansible is running on Fedora 32, target system was Debian Buster.

##### STEPS TO REPRODUCE
On apt system run

    $ ansible-playbook playbook.yaml

where *playbook.yaml* contains:

```yaml
- hosts: localhost
  become: true
  tasks:
    - apt_key:
        url: https://read-token:@packagecloud.io/repo/channel/gpgkey
```

##### EXPECTED RESULTS
The apt_key module should download the key and add it to the system.

##### ACTUAL RESULTS
The key file is not downloaded.

```paste below
TASK [sensu.sensu_go.install : Add apt key] *************************************************************************                                                                                                                         
task path: /vagrant/ansible_collections/sensu/sensu_go/roles/install/tasks/apt/prepare.yml:17                                                                                                                                                 
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: vagrant                                                                                                                                                                                      
<127.0.0.1> EXEC /bin/sh -c 'echo ~vagrant && sleep 0'                                                                                                                                                                                        
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/vagrant/.ansible/tmp `"&& mkdir /home/vagrant/.ansible/tmp/ansible-tmp-1596123703.15-7698-147352890428496 && echo ansible-tmp-1596123703.15-7698-147352890428496="` echo /home/vagrant/.ansible/tmp/ansible-tmp-1596123703.15-7698-147352890428496 `" ) && sleep 0'                                                                                                                                                       
Using module file /home/vagrant/venv/local/lib/python2.7/site-packages/ansible/modules/packaging/os/apt_key.py                                                                                                                                
<127.0.0.1> PUT /home/vagrant/.ansible/tmp/ansible-local-73130sQlx2/tmpTka0Og TO /home/vagrant/.ansible/tmp/ansible-tmp-1596123703.15-7698-147352890428496/AnsiballZ_apt_key.py                                                               
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/vagrant/.ansible/tmp/ansible-tmp-1596123703.15-7698-147352890428496/ /home/vagrant/.ansible/tmp/ansible-tmp-1596123703.15-7698-147352890428496/AnsiballZ_apt_key.py && sleep 0'                  
<127.0.0.1> EXEC /bin/sh -c 'sudo -H -S -n  -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-arcvvjrkaonnmhuagbribxaggrraarss ; /home/vagrant/venv/bin/python2 /home/vagrant/.ansible/tmp/ansible-tmp-1596123703.15-7698-147352890428496/AnsiballZ_apt_key.py'"'"' && sleep 0'                                
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/vagrant/.ansible/tmp/ansible-tmp-1596123703.15-7698-147352890428496/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {                           
    "changed": false,                                      
    "invocation": {                                        
        "module_args": {                                   
            "data": null,
            "file": null,                                  
            "id": null,                                    
            "key": null,                                   
            "keyring": null,                               
            "keyserver": null,                             
            "state": "present",                            
            "url": "https://<redacted>:@packagecloud.io/sensu/release-candidate/gpgkey", 
            "validate_certs": true                         
        }                                                  
    },                                                     
    "msg": "Failed to download key at https://<redacted>:@packagecloud.io/sensu/release-candidate/gpgkey: HTTP Error 401: Unauthorized"
}
```

##### ADDITIONAL INFO

Curlng the url prints:

```
$ curl -v https://<redacted>:@packagecloud.io/sensu/release-candidate/gpgkey
...
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
< HTTP/2 302 
< date: Thu, 30 Jul 2020 15:49:16 GMT
< content-type: text/html;charset=utf-8
< content-length: 0
< location: https://d18dy6y14fq314.cloudfront.net/188/13711/gpg/sensu-release-candidate-44E6C781E1E7FF69.pub.gpg?t=<redacted>
...
```

If we add the `-L` switch to `curl`, the file is downloaded just fine. I can also get module to download the file if I force the basic authentication in the `ansible.module_utils.urls.Request.open(...)` call, but apt_key module does not have an argument that would control this parameter.

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/apt_key.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/apt_key.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sivel** [comment]
To clarify some of the issues here:

1. `apt_key` doesn't provide any means to authenticate against the URL
1. If the user/pass are provided in the URL, and a 30x redirect happens, the user/pass are effectively discarded, as the redirect doesn't contain that information.

The typical recommendation is to call `get_url`, `uri`, or `lookup('url')` and use that data to then call `apt_key`. `apt_key` supports both `data` and `file` for key information, that can receive the output from any of the recommended HTTP request plugins.

**Comment by tadeboro** [comment]
> 1. `apt_key` doesn't provide any means to authenticate against the URL

This is not entirely true because http://user:pass@domain.com is a way to pass credentials to the module. There are no parameters to set credentials explicitly, but adding them to the URL is exactly the same as if they were passed in via the usual url_username and url_password params.

> 2. If the user/pass are provided in the URL, and a 30x redirect happens, the user/pass are effectively discarded, as the redirect doesn't contain that information.

This is true.

> The typical recommendation is to call `get_url`, `uri`, or `lookup('url')` and use that data to then call `apt_key`. `apt_key` supports both `data` and `file` for key information, that can receive the output from any of the recommended HTTP request plugins.

This is the way I worked around this issue: I used the uri module to fetch the key and then fed the result to the apt_key module. But even when using the uri module, I need to pass the `force_basic_auth: true` parameter or the exact same thing happens: authentication data is not passed to the redirected request.

This confused me a bit because it behaves differently from the tools I am used to (curl, wget mainly).

**Comment by tadeboro** [comment]
Just for the sake of completeness, this is the "transformation" I did in our role to get things working. I sterted with:

```
- apt_key:
    url: user:pass@some.url
```

and ended up with this:

```
- uri:
    url: user:pass@some.url
    force_basic_auth: true
    return_content: true
  register: apt_key_download
  changed_when: false
  check_mode: false

- apt_key:
    data: "{{ apt_key_download.content }}"
```

The fix for my immediate problem was the addition `force_basic_auth: true`. All other cruft (`changed_when`, `check_mode`) is there just to ensure that the change detection and check mode are working correctly (this is how apt_key behaves when the download is not problematic).

All that being said, I do not see what would be the best way of fixing the issue here. If I put myself in maintainer shoes, I would deprecate the `url` parameter and instruct users to fetch the key using one of the other modules. Less code to maintain and the module becomes more focused.

But if I look at the transformation I had to make from the end-users perspective, I would much rather see a few additional parameters on the apt_key module that allow me to "tweak" the download process. Because then I do not have to worry about all that check mode and idempotency stuff.

I can prepare a fix for this issue, but I would rather not spend my time implementing the "wrong" thing.
